tests/devapps/MauiApps/MauiAppB2C/Features/LogOn/B2CAuthenticationService.cs

```diff
@@ -0,0 +1,175 @@
+                                                        .WithParentActivityOrWindow(parentWindow)
+                                                        .ExecuteAsync()
+                                                        .ConfigureAwait(false);
+
+            var newContext = UpdateUserInfo(authResult);
+            return newContext;
+        }
+
+        public async Task<UserContext> SignOutAsync()
+        {
+
+            IEnumerable<IAccount> accounts = await _pca.GetAccountsAsync().ConfigureAwait(false);
+            while (accounts.Any())
+            {
+                await _pca.RemoveAsync(accounts.FirstOrDefault()).ConfigureAwait(false);
+                accounts = await _pca.GetAccountsAsync().ConfigureAwait(false);
+            }
+            var signedOutContext = new UserContext();
+            signedOutContext.IsLoggedOn = false;
+            return signedOutContext;
+        }
+
+        private IAccount GetAccountByPolicy(IEnumerable<IAccount> accounts, string policy)
+        {
+            foreach (var account in accounts)
+            {
+                string userIdentifier = account.HomeAccountId.ObjectId.Split('.')[0];
+                if (userIdentifier.EndsWith(policy.ToLower())) return account;
+            }
+
+            return null;
+        }
+
+        private string Base64UrlDecode(string s)
+        {
+            s = s.Replace('-', '+').Replace('_', '/');
+            s = s.PadRight(s.Length + (4 - s.Length % 4) % 4, '=');
+            var byteArray = Convert.FromBase64String(s);
+            var decoded = Encoding.UTF8.GetString(byteArray, 0, byteArray.Count());
+            return decoded;
+        }
+
+        public UserContext UpdateUserInfo(AuthenticationResult ar)
+        {
+            var newContext = new UserContext();
+            newContext.IsLoggedOn = false;
+            JObject user = ParseIdToken(ar.IdToken);
+
+            newContext.AccessToken = ar.AccessToken;
+            newContext.Name = user["name"]?.ToString();
+            newContext.UserIdentifier = user["oid"]?.ToString();
+
+            newContext.GivenName = user["given_name"]?.ToString();
+            newContext.FamilyName = user["family_name"]?.ToString();
+
+            newContext.StreetAddress = user["streetAddress"]?.ToString();
+            newContext.City = user["city"]?.ToString();
+            newContext.Province = user["state"]?.ToString();
+            newContext.PostalCode = user["postalCode"]?.ToString();
+            newContext.Country = user["country"]?.ToString();
+
+            newContext.JobTitle = user["jobTitle"]?.ToString();
+
+            var emails = user["emails"] as JArray;
+            if (emails != null)
+            {
+                newContext.EmailAddress = emails[0].ToString();
+            }
+            newContext.IsLoggedOn = true;
+
+            return newContext;
+        }
+
+        JObject ParseIdToken(string idToken)
```

Comment:
Please don't do this. AuthenticationResult has a ClaimsPrincipal which is the exactly the parsed id token.